aioasuswrt/connection.py

```diff
@@ -95,30 +95,34 @@ def __init__(self, host, port, username, password):
     async def async_run_command(self, command, first_try=True):
         """Run a command through a Telnet connection. If first_try is True a second
         attempt will be done if the first try fails."""
-        if not self.is_connected:
-            await self.async_connect()
 
-        # We can send the command
-        try:
-            async with self._io_lock:
+        async with self._io_lock:
+            try:
+                if not self.is_connected:
+                    await self._async_connect()
                 # Let's add the path and send the command
                 full_cmd = f"{_PATH_EXPORT_COMMAND} && {command}"
                 self._writer.write((full_cmd + "\n").encode("ascii"))
                 # And read back the data till the prompt string
                 data = await asyncio.wait_for(
                     self._reader.readuntil(self._prompt_string), 9
                 )
-        except (BrokenPipeError, LimitOverrunError, IncompleteReadError):
-            # Writing has failed, Let's close and retry if necessary
-            self.disconnect()
-            if first_try:
-                return await self.async_run_command(command, False)
-            else:
-                _LOGGER.warning("connection is lost to host.")
-                return []
-        except TimeoutError:
-            _LOGGER.error("Host timeout.")
-            return []
+            except (BrokenPipeError, LimitOverrunError, IncompleteReadError):
+                # Writing has failed, Let's close and retry if necessary
+                self.disconnect()
+                if first_try:
+                    return await self.async_run_command(command, False)
```

Comment:
Ahh approved it a bit too quickly, so calls with locks are of course always difficult, but now we have a deadlock here, as this call will again try to obtain the lock. (also visible in the tests, which now deadlock).

So my suggestion would be to do the disconnect in both `except` clauses, and set some error string variable (which is set to empty at the start of the method).

Than outside the `async with self._io_lock` the value of the error string can be checked. If not empty log the value, and if it's the first_try redo the command. This should also refactor some duplicate code from this method.

Reply:
Wops, sorry for that. Now it should be fine. Tested this on my local machine this time :D